File: 2023/01/02/myHashSet.js
/**
 * https://leetcode.com/problems/design-hashset/
 * @author Andrew Jarombek
 * @since 1/2/2023
 */

const MyHashSet = function() {
    this.buckets = [];

    for (let i = 0; i < 100; i++)
        this.buckets.push([]);
};

/**
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.add = function(key) {
    this.buckets[(key * 31) % 100].push(key);
};

/**
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.remove = function(key) {
    let bucket = this.buckets[(key * 31) % 100];
    this.buckets[(key * 31) % 100] = bucket.filter((val) => val !== key);
};

/**
 * @param {number} key
 * @return {boolean}
 */
MyHashSet.prototype.contains = function(key) {
    return this.buckets[(key * 31) % 100].includes(key);
};

const hashSet = new MyHashSet();
hashSet.add(1);
hashSet.add(2);

console.assert(hashSet.contains(1));
console.assert(!hashSet.contains(3));

hashSet.add(2);
console.assert(hashSet.contains(2));

hashSet.remove(2);
console.assert(!hashSet.contains(2));

File: 2023/04/24/java/src/main/java/com/jarombek/MyHashSet.java
/*
 * https://leetcode.com/problems/design-hashset/
 * @author Andrew Jarombek
 * @since 4/24/2023
 */

package com.jarombek;

import java.util.ArrayList;
import java.util.List;

public class MyHashSet {
    private List<Integer>[] buckets;

    public MyHashSet() {
        this.buckets = new List[2000];
    }

    public void add(int key) {
        if (!this.contains(key)) {
            int index = key * 31 % 2000;
            List<Integer> bucket = this.buckets[index];

            if (bucket == null) {
                bucket = new ArrayList<>();
                this.buckets[index] = bucket;
            }

            bucket.add(key);
        }
    }

    public void remove(int key) {
        int index = key * 31 % 2000;
        List<Integer> bucket = this.buckets[index];

        if (bucket != null) {
            bucket.remove(Integer.valueOf(key));
        }
    }

    public boolean contains(int key) {
        int index = key * 31 % 2000;
        List<Integer> bucket = this.buckets[index];

        if (bucket == null) {
            return false;
        }

        for (int value : bucket) {
            if (value == key) {
                return true;
            }
        }

        return false;
    }
}
